lib/Dialect/Triton/Transforms/RewriteTMAToLegacy.cpp

```diff
@@ -0,0 +1,428 @@
+  Value mask;
+  for (auto i : boundaryCheck.value_or(check_all)) {
+    auto offsetWithRange =
+        getExpandedOffsetWithRange(builder, loc, block_shape, offsets, i);
+
+    // Compare with lower bound
+    Value lowerBound = builder.create<mlir::arith::ConstantIntOp>(
+        loc, 0, builder.getI64Type());
+    Value splatLowerBound = builder.create<triton::SplatOp>(
+        loc, offsetWithRange.getType(), lowerBound);
+    Value cmpLower = builder.create<arith::CmpIOp>(
+        loc, arith::CmpIPredicate::sge, offsetWithRange, splatLowerBound);
+
+    // Compare with upper bound
+    Value splatUpperBound = builder.create<triton::SplatOp>(
+        loc, offsetWithRange.getType(), shape[i]);
+    Value cmpUpper = builder.create<arith::CmpIOp>(
+        loc, arith::CmpIPredicate::slt, offsetWithRange, splatUpperBound);
+
+    // And and broadcast
+    Value andResult = builder.create<arith::AndIOp>(loc, cmpLower, cmpUpper);
+    Value broadcasted =
+        builder.create<triton::BroadcastOp>(loc, maskTensorType, andResult);
+
+    // And up all results
+    if (!mask) {
+      mask = broadcasted;
+    } else {
+      mask = builder.create<arith::AndIOp>(loc, mask, broadcasted);
+    }
+  }
+
+  return mask;
+}
+
+// Note this has been adapted from RewriteTensorPointer.cpp. It appears
+// to be getting the values used for the masked out elements
+Value generateOther(OpBuilder &builder, const Location &loc,
+                    const std::optional<triton::PaddingOption> &padding,
+                    Value base, ArrayRef<std::int64_t> block_shape) {
+  // Create element attribute
+  auto elementType = cast<triton::PointerType>(base.getType()).getPointeeType();
+  auto otherTensorType = RankedTensorType::get(block_shape, elementType);
+
+  // Set zero padding value (the default)
+  TypedAttr attr = builder.getZeroAttr(elementType);
+
+  // Float NaN padding case
+  if (padding == triton::PaddingOption::PAD_NAN) {
+    assert(!elementType.isIntOrIndex());
+    auto apNaN =
+        llvm::APFloat::getNaN(cast<FloatAttr>(attr).getValue().getSemantics());
+    attr = builder.getFloatAttr(elementType, apNaN);
+  }
+
+  // Create tensor
+  Value constant = builder.create<arith::ConstantOp>(loc, attr);
+  return builder.create<triton::SplatOp>(loc, otherTensorType, constant);
+}
+
+// Note the attributes below don't normally appear on the TMA load/store ops.
+// However, if we are lowering block pointers these will be added as addition
+// attributes on the TMA operation for us to retrieve.
+std::optional<ArrayRef<int>> getBoundaryCheck(Operation *op) {
+  std::string_view attr_name = "boundaryCheck";
+  if (auto attr = op->getAttrOfType<DenseI32ArrayAttr>(attr_name)) {
+    return attr.asArrayRef();
+  }
+  return std::nullopt;
+}
+std::optional<triton::PaddingOption> getPadding(Operation *op) {
+  std::string_view attr_name = "padding";
+  if (auto attr = op->getAttrOfType<triton::PaddingOptionAttr>(attr_name)) {
```

Comment:
I don't think we want to keep these since the attributes aren't used anywhere and we will be getting rid of block pointers.